src/lib/y2partitioner/device_graphs.rb

```diff
@@ -0,0 +1,47 @@
+require "singleton"
+
+module Y2Partitioner
+  # Object that holds and manipulates of storage device graphs.
+  class DeviceGraphs
```

Comment:
I must be overlooking something and, as a result, oversimplifying the whole thing. But, cannot this be as simple as this?
```ruby
module Y2Partitioner
  class Devicegraphs
    # Devicegraph representing the system
    attr_reader :system
    # Working devicegraph, to be directly modified during the partitioner execution
    attr_reader :current

    def initialize(system: nil, initial: nil)
      @system = system || Y2Storage::StorageManager.instance.probed
      initial ||= Y2Storage::StorageManager.instance.staging
      @current = initial.dup
    end 

    def transaction(&block)
      back = self.current.dup
      res = block.call
      self.current = back unless res 
      res 
    end 

    class << self
      # Returns the singleton instance.
      def instance
        create_instance unless @instance
        @instance
      end 

      # Creates the singleton instance with customized devicegraphs. To be used
      # during the initialization of the partitioner.
      def create_instance(system, initial)
        @instance = new(system: system, initial: initial)
      end 

      # Make sure only .instance and .create_instance can be used to create objects
      private :new, :allocate
    end 
  end 
end
```

Then, we could initialize the partitioner with something like:
```ruby
manager = Y2Storage::StorageManager.instance
initial = from_scratch? ? manager.probed : manager.staging
Devicegraphs.create_instance(manager.probed, initial)
```

Most widget would then work with the passed devices without caring much about the devicegraphs. For example, they can do
```ruby
the_partition.create_filesystem(131)
```
which will modify `current`, although the widget does not need to know, it simply modifies the passed `the_partition`.

For the infamous multi-step widgets, they will do
```ruby
Devicegraphs.transaction do
   # Directly modify the objects
   # Return false to discard the changes
end
```

That's everything we needed. Isn't it?
And is deadly simple compared to this class, IMHO.

@teclator, @mvidner How do you see it? What am I overlooking that enforces such a complex class?